src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+        """
+        Returns the list of labels associated to the trained model.
+        """
+        return self.labels
+
+    def _get_loss_and_metric(self):
+        """
+        Use the loss corresponding to the loss_name field.
+        """
+        if self.loss_name == "mean_squared_error":
+            self.loss = tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.NONE)
+        elif self.loss_name == "sparse_categorical_crossentropy":
+            self.loss = tf.keras.losses.SparseCategoricalCrossentropy(reduction=tf.keras.losses.Reduction.NONE, from_logits=True)
+        elif self.loss_name == "qa_loss":
+            self.loss = QALoss()
+        else:
+            raise ValueError(
+                "Allowed losses are mean_squared_error or sparse_categorical_crossentropy. Given: {}".format(self.loss_name)
+            )
+
+        if self.metric_name == "sparse_categorical_accuracy":
+            self.train_acc_metric = tf.keras.metrics.SparseCategoricalAccuracy(name='train_accuracy')
+            self.test_acc_metric = tf.keras.metrics.SparseCategoricalAccuracy(name='test_accuracy')
+        elif self.metric_name == "qa_metric":
+            pass
+        else:
+            raise ValueError(
+                "Allowed metrics are sparse_categorical_accuracy and qa_metric. Given: {}".format(self.metric_name)
+            )
+
+        self.test_loss_metric = tf.keras.metrics.Mean(name='test_loss')
+
+    def _create_summary_writer(self, log_path):
+        """
+        Create a summary writer to be able to read the logs in Tensorboard.
+        """
+        self.log_path = log_path
+        self.train_writer = tf.summary.create_file_writer(log_path + "/train")
+        self.test_writer = tf.summary.create_file_writer(log_path + "/test")
+
+    @abstractmethod
+    def _create_features(self, tokenizer, data_dir):
+        """
+        Create the features for the training and validation data.
+        Args:
+          tokenizer: the tokenizer used for encoding the textual data into features.
+          data_dir: the directory path where the data are.
+          cache_dir: the directory path where the cached data are.
+        """
+        pass
+
+    @abstractmethod
+    def _load_cache(self, cached_file):
+        """
+        Load a cached TFRecords dataset.
+        Args:
+          cached_file: the TFRecords file path.
+        """
+        pass
+
+    @abstractmethod
+    def _save_cache(self, mode, cached_file):
+        """
+        Save a cached TFRecords dataset.
+        Args:
+          mode: the dataset to be cached.
+          cached_file: the file path where the TFRecords will be saved.
+        """
+        pass
+
+    def _preprocessing_data(self, tokenizer, cache_dir):
+        """
+        Preprocess the training and validation data.
+        Args:
+          tokenizer: the tokenizer used for encoding the textual data into features.
+          data_dir: the directory path where the data are.
```

Comment:
I do agree, but manage both in same time was too confusing for me so I decided to go one at a time, first I implement for the files and once it will be ok, I will go for the streamed data.